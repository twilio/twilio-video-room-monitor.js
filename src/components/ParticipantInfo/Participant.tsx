import React from 'react';
import { Accordion } from '../typography/Accordion/Accordion';
import Datum from '../typography/Datum/Datum';
import {
  Participant as ParticipantImpl,
  LocalVideoTrackPublication,
  RemoteVideoTrackPublication,
  LocalAudioTrackPublication,
  RemoteAudioTrackPublication,
} from 'twilio-video';
import useParticipantNetworkQualityLevel from '../../hooks/useParticipantNetworkQualityLevel/useParticipantNetworkQualityLevel';
import useParticipantIsReconnecting from '../../hooks/useParticipantIsReconnecting/useParticipantIsReconnecting';
import usePublications from '../../hooks/usePublications/usePublications';

import { VideoTrackPublicationInfo } from './VideoTrackPublicationInfo';
import { AudioTrackPublicationInfo } from './AudioTrackPublicationInfo';

export const Participant = ({ participant }: { participant: ParticipantImpl }) => {
  const networkQualityLevel = useParticipantNetworkQualityLevel(participant);
  const isReconnecting = useParticipantIsReconnecting(participant);

  const publications = usePublications(participant);

  const dataTrackPublications = publications.filter((publication) => publication.kind === 'data');
  const audioTrackPublications = publications.filter((publication) => publication.kind === 'audio');
  const videoTrackPublications = publications.filter((publication) => publication.kind === 'video');

  return (
    <Accordion label={participant.identity}>
      <Datum label="SID" value={participant.sid} />
      <Datum label="isReconnecting" value={String(isReconnecting)} />
      <Datum label="networkQualityLevel" value={String(networkQualityLevel)} />
      <Accordion label={`Data Tracks (${dataTrackPublications.length})`}>{/* TODO */}</Accordion>
      <Accordion label={`Audio Tracks (${audioTrackPublications.length})`}>
        {audioTrackPublications.map((audioTrackPublication) => (
          <AudioTrackPublicationInfo
            key={audioTrackPublication.trackSid}
            publication={audioTrackPublication as LocalAudioTrackPublication | RemoteAudioTrackPublication}
          />
        ))}
      </Accordion>
      <Accordion label={`Video Tracks (${videoTrackPublications.length})`}>
        {videoTrackPublications.map((videoTrackPublication) => (
          <VideoTrackPublicationInfo
            key={videoTrackPublication.trackSid}
            publication={videoTrackPublication as LocalVideoTrackPublication | RemoteVideoTrackPublication}
          />
        ))}
      </Accordion>
    </Accordion>
  );
};
