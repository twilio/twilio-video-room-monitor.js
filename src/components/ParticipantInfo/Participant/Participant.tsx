import React from 'react';
import { Accordion } from '../../typography/common/Accordion/Accordion';
import Datum from '../../typography/common/Datum/Datum';
import {
  Participant as ParticipantImpl,
  LocalVideoTrackPublication,
  RemoteVideoTrackPublication,
  LocalAudioTrackPublication,
  RemoteAudioTrackPublication,
  LocalDataTrackPublication,
  RemoteDataTrackPublication,
} from 'twilio-video';
import useParticipantNetworkQualityLevel from '../../../hooks/useParticipantNetworkQualityLevel/useParticipantNetworkQualityLevel';
import useParticipantIsReconnecting from '../../../hooks/useParticipantIsReconnecting/useParticipantIsReconnecting';
import usePublications from '../../../hooks/usePublications/usePublications';

import { VideoTrackPublicationInfo } from '../VideoTrackPublicationInfo/VideoTrackPublicationInfo';
import { AudioTrackPublicationInfo } from '../AudioTrackPublicationInfo/AudioTrackPublicationInfo';
import { DataTrackPublicationInfo } from '../DataTrackPublicationInfo/DataTrackPublicationInfo';

export const Participant = ({ participant, isLocal }: { participant: ParticipantImpl; isLocal?: boolean }) => {
  const networkQualityLevel = useParticipantNetworkQualityLevel(participant);
  const isReconnecting = useParticipantIsReconnecting(participant);

  const publications = usePublications(participant, isLocal);

  const dataTrackPublications = publications.filter((publication) => publication.kind === 'data');
  const audioTrackPublications = publications.filter((publication) => publication.kind === 'audio');
  const videoTrackPublications = publications.filter((publication) => publication.kind === 'video');

  return (
    <Accordion label={participant.identity}>
      <Datum label="SID" value={participant.sid} />
      <Datum label="isReconnecting" value={isReconnecting} />
      <Datum label="networkQualityLevel" value={networkQualityLevel} />
      <Accordion label={`Data Tracks (${dataTrackPublications.length})`}>
        {dataTrackPublications.map((dataTrackPublication) => (
          <DataTrackPublicationInfo
            key={dataTrackPublication.trackSid}
            publication={dataTrackPublication as LocalDataTrackPublication | RemoteDataTrackPublication}
          />
        ))}
      </Accordion>
      <Accordion label={`Audio Tracks (${audioTrackPublications.length})`}>
        {audioTrackPublications.map((audioTrackPublication) => (
          <AudioTrackPublicationInfo
            key={audioTrackPublication.trackSid}
            publication={audioTrackPublication as LocalAudioTrackPublication | RemoteAudioTrackPublication}
          />
        ))}
      </Accordion>
      <Accordion label={`Video Tracks (${videoTrackPublications.length})`}>
        {videoTrackPublications.map((videoTrackPublication) => (
          <VideoTrackPublicationInfo
            key={videoTrackPublication.trackSid}
            publication={videoTrackPublication as LocalVideoTrackPublication | RemoteVideoTrackPublication}
          />
        ))}
      </Accordion>
    </Accordion>
  );
};
