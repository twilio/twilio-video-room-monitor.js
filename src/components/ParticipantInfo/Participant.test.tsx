import React from 'react';
import { shallow } from 'enzyme';
import { Participant } from './Participant';

jest.mock('../../hooks/useParticipantIsReconnecting/useParticipantIsReconnecting', () => () => false);
jest.mock('../../hooks/useParticipantNetworkQualityLevel/useParticipantNetworkQualityLevel', () => () => 4);
jest.mock('../../hooks/usePublications/usePublications', () => () => [
  { trackName: 'camera-123456', kind: 'video', trackSid: 'testVideoSid' },
  { trackName: 'microphone-123456', kind: 'audio', trackSid: 'testAudioSid' },
]);

describe('the Participant component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Participant participant={{ identity: 'mockIdentity', sid: 'testSid' } as any} />);
    expect(wrapper).toMatchSnapshot();
  });
});
