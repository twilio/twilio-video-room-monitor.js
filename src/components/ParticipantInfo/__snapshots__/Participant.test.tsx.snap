// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Participant component should render correctly 1`] = `
<Accordion
  label="mockIdentity"
>
  <Memo(Datum)
    label="SID"
    value="testSid"
  />
  <Memo(Datum)
    label="isReconnecting"
    value="false"
  />
  <Memo(Datum)
    label="networkQualityLevel"
    value="4"
  />
  <Accordion
    label="Data Tracks (0)"
  />
  <Accordion
    label="Audio Tracks (1)"
  >
    <AudioTrackPublicationInfo
      key="testAudioSid"
      publication={
        Object {
          "kind": "audio",
          "trackName": "microphone-123456",
          "trackSid": "testAudioSid",
        }
      }
    />
  </Accordion>
  <Accordion
    label="Video Tracks (1)"
  >
    <VideoTrackPublicationInfo
      key="testVideoSid"
      publication={
        Object {
          "kind": "video",
          "trackName": "camera-123456",
          "trackSid": "testVideoSid",
        }
      }
    />
  </Accordion>
</Accordion>
`;
